src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets

```diff
@@ -161,4 +161,26 @@
 
     <CallTarget Targets="$(TargetsTriggeredByCompilation)" Condition="'$(TargetsTriggeredByCompilation)' != ''" />
   </Target>
+
+  <!-- When we load a project in Visual Studio, the project system first does an evaluation pass of the
+       project, and hands the resulting list of <Compile> items to the language service. It then does an
+       execution pass executing CoreCompile passing SkipCompilerExecution=true and ProvideCommandLineArgs=true,
+       that resulting command line string is where we get our compiler switches. The execution pass is much
+       slower than the evaluation pass, so there's a window of time where we have a list of files, but not
+       any options yet.
+
+       Because there's a gap, that means there's a time where we are parsing source files with the default
+       parse options. We'll then have to reparse them a second time which isn't great. It also means any
+       cache lookups we do won't have the right options either, so the cache lookups might miss.
+
+       To help this, we'll have a property for the evaluation pass which is an "approximation" of the
+       options that would come out of CoreCompile, but only the ones that matter for parsing. It's acceptable
+       for this to be imperfect: once the execution pass is complete we'll use those options instead,
+       so any behaviors here that don't match the real command line generation will only be temporary, and
+       probably won't be any worse than having no options at all. -->
+       
+  <PropertyGroup>
+    <CommandLineArgsForDesignTimeEvaluation>-langversion:$(LangVersion)</CommandLineArgsForDesignTimeEvaluation>
+    <CommandLineArgsForDesignTimeEvaluation Condition="'$(DefineConstants)' != ''">$(CommandLineArgsForDesignTimeEvaluation) -define:$(DefineConstants)</CommandLineArgsForDesignTimeEvaluation>
```

Comment:
One issue that was identified here was the SDK computes some of its constants during execution, which means this optimization doesn't always work as ideally as we'd like. There's other problems with that and we'll investigate further in what it'd take to do. We have some recently added optimizations around avoiding reparsing if there's only a preprocessor directive change and the file doesn't have any preprocessor directives; having us at least get the langversion over means we have a better chance of that optimization being able to kick in.

tl;dr; this doesn't work in all cases, but that's better than working on no cases.

Reply:
just curious, is there any telemetry for when "misses" occur here? something we could look at occasionally and say--oops, we added this option which affects parsing, we can see that we need to update this target.

Reply:
We do have telemetry that we added when we initially thought of doing this: https://github.com/dotnet/roslyn/blob/c756bbabff2344cc1813c968245d8477a9ae3673/src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs#L280. My concern isn't the lack of the telemetry, but how we'll know to actually look at it if/when the time comes. I admit I don't have any good ideas there.